test: Add GraphQL conformance tests

karatakis opened the pull request:
**Summary:**  
Use [GraphQ spec](https://spec.graphql.org/October2021/) and write conformance tests for Tailcall.

**Build & Testing:**

- [X] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [X] I have added relevant unit & integration tests.
- [X] I have updated the [documentation] accordingly.
- [X] I have performed a self-review of my code.
- [X] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

meskill commented on tests/core/snapshots/graphql-conformance-002.md_0.snap:
```diff
@@ -0,0 +1,39 @@
+    "content-type": "application/json"
+  },
+  "body": {
+    "data": {
```
instead of recreating right snapshots here we can mark failed tests as skip with relevant comment. And for that test we can create an issue to fix this and enable test

karatakis replied:
That is a good strategy to know which stuff are not currently implemented

meskill commented on tests/execution/graphql-conformance-004.md:
```diff
@@ -0,0 +1,49 @@
+        user(id: 4) {
+          id
+          name
+          smallPic: profilePic(size: 64)
+          bigPic: profilePic(size: 1024)
```
this test is failing because aliases are not handled at all at selection_set for the graphql upstream. It applies both for async_graphql and jit implementation

karatakis replied:
Noted.

meskill commented on tests/execution/graphql-conformance-005.md:
```diff
@@ -0,0 +1,46 @@
+        customer: user(id: 4) {
+          id
+          name
+          date_of_birth: dob
```
seems like it fails also because of the aliases not working as in the previous test

karatakis replied:
Correct

meskill commented on tests/execution/graphql-conformance-006.md:
```diff
@@ -0,0 +1,177 @@
+      query withFragments {
+        user(id: 4) {
+          friends(first: 10) {
+            ...friendFields
```
fragments are not supported in this moment for graphql upstream. I'm not even sure if we can do this propely with current async_graphql api. For jit we need to desing actually how to handle fragments for the graphql upstream

karatakis replied:
There is no need to support it at the moment. I just documented the behavior so we can track the functionality that is missing

meskill commented on tests/execution/graphql-conformance-007.md:
```diff
@@ -0,0 +1,83 @@
+      query FragmentTyping {
+        profiles(handles: ["user-1", "user-2"]) {
+          handle
+          ...userFragment
+          ...pageFragment
```
this also failing because of the fragments I believe

karatakis replied:
Yeah, we don't send them over to the target graphql server. Maybe we might need an issue to fix that

karatakis commented on tests/execution/graphql-conformance-001.md:
````diff
@@ -0,0 +1,104 @@
+```graphql @config
````
Noted, I added names that describe what is the test about on each file

karatakis commented on tests/execution/graphql-conformance-002.md:
```diff
@@ -0,0 +1,76 @@
+  name: String!
+  city: String
+  birthday: BirthDay!
+  friends: [User!]!
```
Yeah, it works okay for JIT, I tested it

meskill commented:
I think we still missing some tests for some graphql features (correct me if I'm wrong):
- use interfaces and unions as fragment `on` type
- default values for arguments including default values for nested fields in input type
- validation tests for negative scenarios. So basically for every feature we're testing we should have several negative cases scenarios

karatakis commented:
@meskill 

- fragments on unions are implemented
- default input values are tested
- All negative paths are failing because we have validation disabled. So the current behavior is ignoring the field altogether

meskill commented on tests/execution/graphql-conformance-012.md:
```diff
@@ -0,0 +1,164 @@
+        height
+        ...metaFragment
+      }
+      fragment metaFragment on photo {
+        meta {
+          iso
+        }
+      }
```
is this wrong fragment?

meskill commented on tests/execution/graphql-conformance-012.md:
```diff
@@ -0,0 +1,164 @@
+      }
+
+# Negative: missing fragment
+# - method: POST
```
why comment them?

this test is skipped anyway

meskill commented:
> @meskill
> 
> * fragments on unions are implemented
> * default input values are tested
> * All negative paths are failing because we have validation disabled. So the current behavior is ignoring the field altogether

- what about fragments on interfaces and the union itself? I think we can specify fragments not only on concrete type but on the interface and union itself
- I see default inputs but input could be also a complex object that could contain default fields
- what validation you're talking about? Many validations comes from async_graphql out of the box, like required fields, query validation against schema etc.

meskill commented:
I found additional case missing - list inside lists. See https://github.com/graphql/graphql-spec/issues/749. It doesn't work in both implementations

karatakis commented:
@meskill 
- added fragment example (it fails, so it is disabled)
- complex input with optional is achieved
- implemented many negative scenarios (but because we have the validation from async_graphql disabled for performance reasons, they are failing)
- List of lists example is implemented

This is an excellent start to fix many issues regarding the complete GraphQL spec. There is still room for improvement, but the tests highlight the rough edges.

karatakis commented:
@meskill 
- added tests for an input field with an array
- added test for an input type that defines default values
- added descriptions of why tests are disabled

meskill commented on tests/execution/graphql-conformance-002.md:
```diff
@@ -0,0 +1,109 @@
+skip: true
+---
+
+# Test complex nested query. Skipped because Tailcall does not send the whole query to the remote server. It sends a shallow version of the query.
```
```suggestion
# Test complex nested query. 

TODO: Skipped because Tailcall does not send the whole query to the remote server. It sends a shallow version of the query.
```

Maybe format like this? In my opinion it's more visible and allows to provide more info

meskill commented on tests/execution/graphql-conformance-http-002.md:
```diff
@@ -0,0 +1,103 @@
+      }
+
+# Negative: invalid selection at nested
+# Disabled because async_graphql::dynamic does not perform validation
```
```suggestion
# TODO: Disabled because async_graphql::dynamic does not perform validation
```

also to increase visibility

meskill commented on tests/execution/graphql-conformance-http-007.md:
```diff
@@ -0,0 +1,84 @@
+skip: true
+---
+
+# Test named fragments. Skipped because Tailcall does not construct correctly the interface types based on \_\_typename field.
```
I think it's not working because we not using our discriminator https://github.com/tailcallhq/tailcall/blob/f464794bd32abe3224eb96e1de09d6fccd766148/src/core/blueprint/union_resolver.rs#L34 for interfaces and only for unions. The upstreams in most cases won't provide `__typename` info for us anyway and we need to figure out the type on our side, that's the discrininator job

meskill commented on tests/execution/graphql-conformance-http-007.md:
```diff
@@ -0,0 +1,84 @@
+    body:
+      - id: 1
+        handle: user-1
+        __typename: User
```
upstream won't send `__typename` in most cases, especially rest upstream

meskill replied:
could you also check other tests for this? instead of relying on `__typename` from the upstream Tailcall should be able to figure the type on it's own

karatakis replied:
I just removed the __typename. This test can test this behavior

meskill commented on tests/execution/graphql-conformance-http-012.md:
```diff
@@ -0,0 +1,139 @@
+      }
+
+# Negative: missing fragment
+# Disabled because async_graphql::dynamic does not perform validation
```
could you expand on what validation is expected in those cases?

meskill commented on tests/execution/graphql-conformance-015.md:
```diff
@@ -0,0 +1,185 @@
+#         user(id: 4) {
+#           id
+#           name
+#           featuredVideoPreview
```
is this test right?

the field set args as required and we need to provide that arg (or that is validation missed case?) and we need to test that `hdr` is got defaulted to `true`

karatakis replied:
changed it to `featuredVideoPreview(video: VideoSize! = {}): String!`. I set up another graphql implementation and verified that hdr is true. I disable the test for now.

meskill commented on tests/execution/graphql-conformance-http-015.md:
```diff
@@ -0,0 +1,182 @@
+input VideoSize {
+  width: Int!
+  height: Int!
+  hdr: Boolean
```
this test differs from its companion test for graphql because there is no default values inside input type

karatakis replied:
good catch :+1:
